services/src/pro/users/oidc.rs

```diff
@@ -0,0 +1,919 @@
+use openidconnect::core::{CoreAuthDisplay, CoreAuthenticationFlow, CoreAuthPrompt, CoreClaimName, CoreClaimType, CoreClient, CoreClientAuthMethod, CoreGenderClaim, CoreGrantType, CoreJsonWebKey, CoreJsonWebKeyType, CoreJsonWebKeyUse, CoreJweContentEncryptionAlgorithm, CoreJweKeyManagementAlgorithm, CoreJwsSigningAlgorithm, CoreProviderMetadata, CoreResponseMode, CoreResponseType, CoreSubjectIdentifierType, CoreTokenIntrospectionResponse, CoreTokenResponse};
+use serde::{Serialize, Deserialize};
+use url::Url;
+use geoengine_datatypes::primitives::Duration;
+use crate::contexts::Db;
+use crate::error::Error::{IllegalOIDCLoginRequest, OIDCLoginFailed, IllegalOIDCProviderConfig};
+use crate::pro::users::UserId;
+use crate::error::Result;
+use crate::error::Error;
+use crate::pro::util::config::Oidc;
+use crate::pro::util::tests::{SINGLE_NONCE, SINGLE_STATE};
+
+pub type DefaultProviderMetadata = ProviderMetadata<
+    EmptyAdditionalProviderMetadata,
+    CoreAuthDisplay,
+    CoreClientAuthMethod,
+    CoreClaimName,
+    CoreClaimType,
+    CoreGrantType,
+    CoreJweContentEncryptionAlgorithm,
+    CoreJweKeyManagementAlgorithm,
+    CoreJwsSigningAlgorithm,
+    CoreJsonWebKeyType,
+    CoreJsonWebKeyUse,
+    CoreJsonWebKey,
+    CoreResponseMode,
+    CoreResponseType,
+    CoreSubjectIdentifierType
+>;
+
+pub type DefaultJsonWebKeySet = JsonWebKeySet<
+    CoreJwsSigningAlgorithm,
+    CoreJsonWebKeyType,
+    CoreJsonWebKeyUse,
+    CoreJsonWebKey
+>;
+
+pub type DefaultClient = Client<
+    EmptyAdditionalClaims,
+    CoreAuthDisplay,
+    CoreGenderClaim,
+    CoreJweContentEncryptionAlgorithm,
+    CoreJwsSigningAlgorithm, CoreJsonWebKeyType,
+    CoreJsonWebKeyUse,
+    CoreJsonWebKey,
+    CoreAuthPrompt,
+    StandardErrorResponse<BasicErrorResponseType>,
+    CoreTokenResponse,
+    BasicTokenType,
+    CoreTokenIntrospectionResponse,
+    StandardRevocableToken,
+    BasicRevocationErrorResponse
+>;
+
+
+pub struct OIDCRequestsDB {
+    issuer : String,
+    client_id : String,
+    client_secret : String,
+    redirect_uri : String,
+    scopes : Vec<String>,
+    users: Db<HashMap<String, PendingRequest>>,
+    state_function : fn() -> CsrfToken,
+    nonce_function : fn() -> Nonce,
+}
+
+pub struct PendingRequest {
+    nonce : Nonce, //TODO: Is nonce unnecessary in code flow?
+    code_verifier : PkceCodeVerifier,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct AuthCodeRequestURL{
+    url : Url,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct AuthCodeResponse {
+    pub session_state : String,
+    pub code : String,
+    pub state : String,
+}
+
+#[derive(Clone)]
+pub struct ExternalUser {
+    pub id: UserId,
+    pub claims : ExternalUserClaims,
+    pub active: bool,
+}
+
+#[derive(Clone)]
+pub struct ExternalUserClaims {
+    pub external_id: SubjectIdentifier,
+    pub email: String,
+    pub real_name: String,
+}
```

Comment:
Most field are named as in the OIDC / OAuth2.0 standard. Structs are not 1:1, e.g. ExternalUserClaims only contains the ones that Geo Engine requires and is a flat structure rather than a deeply nested one from the standard.

Not sure I understand the camelCase comment. Thought rust was snake case mostly? This matches the field names in the requests/results of OIDC, if that was the question.

Reply:
Yes, so, Geo Engine has an API that uses camelCase.
But when these fields need to be snake_cased because it is part of the OIDC standard, we can leave this as is.
Otherwise, we still use snake_case for the Rust fields but rename them in Serde.

Reply:
I actually had to change the session_state because otherwise the UI linter would complain. 2ed0b38b9f66fcd782833ee250604dcd94de703f